### ISSUE TITLE
Implement a way to perform a backup from a drive

### ISSUE THREAD

**Issue by jviotti** [resin-io/etcher#266]
From https://github.com/resin-io/etcher/issues/265

This option would copy whatever there is on the device as a local `*.img` file, and will compress it as a ZIP archive, in a way that can be easily burnt back with Etcher.

**Comment by jviotti** [comment]
I just had some ideas for this: The problem at the moment is that blindly copying the device contents back means copying a lot of unnecessary data and having a backup image that equals the drive size. If you want to flash that back to a drive that is smaller, but in which the image logically fits, you'll have a hard time figuring out how to reduce appropriately. 

As a solution, we can scan the partition table and only copy what is really necessary.

**Comment by lurch** [comment]
> As a solution, we can scan the partition table and only copy what is really necessary.

I thought I'd already suggested that, but maybe I'd imagined it and it was something I was *going* to suggest, but forgot to :confused: 
Sounds good :+1:

**Comment by jviotti** [comment]
Haha, maybe it was on another related issue and it somehow got lost?

**Comment by alexandrosm** [comment]
@jviotti, reading this again, I think you're making your life difficult. We can start with a "dumb" imaging process, and improve on that later

**Comment by lurch** [comment]
Reading the partition table to determine how much of the disk to _actually_ read, and simply outputting a smaller diskimage at the end of the process, isn't that complicated IMHO :-)  It's only a small part of the process, so I'd see no reason to leave it out.

**Comment by jviotti** [comment]
@alexandrosm The thing is that copying back the whole thing could bring a lot of issues in the case of Etcher. Most users get away with it using something like `dd`, since `dd` will only copy what it can and discard the rest (so doing a backup of the whole thing causes no issues most of the time), however Etcher will complain in that case, which means users can only backup drives that they will re-flash to the same drive or a larger one.

Maybe I'm indeed overthinking it and that's not an usual scenario (maybe people only re-write the backups to the exact same drive most of the time), but if it is, Etcher would feel really awkward.

**Comment by alexandrosm** [comment]
All I'm saying is, let's do the part that's easy now, and solve the hard
part later, if its an issue, rather than block the easy on the hard :)

-- 

*Alexandros Marinos*

Founder & CEO, Resin.io


+1 206-637-5498

@alexandrosm

**Comment by agsdot** [comment]
Any updates/progress on this issue @jviotti?

**Comment by blinkenlight** [comment]
This would be pretty important for me too - I'd just be doing backups and restores (if needed = I screw up) on the same card anyway.

**Comment by jviotti** [comment]
Hey there,

We are sadly not making progress on this one now given that we're preparing some other major features for a new release. This is something we really want to support, so we'll definitely allocate some time for this soon.

That being said, if anyone is motivated enough to make this work, we welcome contributions to https://github.com/resin-io-modules/etcher-image-write, and then to integrate it with the UI (ping me on [Gitter](https://gitter.im/resin-io/etcher) if you're interested and I can provide guidance to implement this)

**Comment by archesys** [comment]
Hi all,

I discover your application some days ago and find it very nice and useful, I work on Linux and a friend on Apple, so I also recommend this application to him.. We work on raspberry so we regularly need to backup and restore sdcard.. The backup function (in zip or tar.gz) would be very useful. no more news on that ? I'm french so if you want me to translate interface in french, I could help you (I don't have so much time but If it's not too big work it may be affordable.

Eric..

**Comment by fzymek** [comment]
+1 for this feature from me as well, very useful for raspberry

**Comment by tomatac** [comment]
This is an essential feature. Currently, what is the recommended way to create an image?

**Comment by jviotti** [comment]
@tomatac You can use `dd` to copy the device data to a file.

**Comment by incyi** [comment]
+1

**Comment by DWatkins** [comment]
+1 

A backup feature would greatly ease my workflow

**Comment by adrianmihalko** [comment]
Still no backup feature?

**Comment by edsonpatricio** [comment]
+1

**Comment by dominikkv** [comment]
+1
Would be great!

**Comment by philomatic** [comment]
Do this!

+1

**Comment by onkelandy** [comment]
Alternative way is to use Apple Disk Utility.. But having a simple way to create an image with only the necessary size would be aaaaawwwwesome!
+1

**Comment by borpin** [comment]
+1 Would be fantastic to have this and not have to go back to dd 👍

**Comment by cgpena** [comment]
+1

**Comment by jarrah31** [comment]
21 months since the original request - I guess this feature isn't going to be implemented?

**Comment by lurch** [comment]
@jarrah31 This is still on the TODO list (and in fact #1671 was a big step towards making this backup feature possible), we've just had lots of other more important features to concentrate on.

**Comment by adammhaile** [comment]
+1

**Comment by xristocodos** [comment]
So even though most people aren't writing the image back to the exact same drive (usb id), they are most certainly moving it to the same SIZE of drive or greater. Just make a disclaimer and fade out the Begin! button (or whatever) if it doesn't detect an adequate size. Cramming images into smaller drives is a dicey tricky artform that I personally don't believe anyone expects Etcher to perform. I'd just put a disclaimer saying that 0.) this feature is somewhat experimental, 1.) the drive must be equal or larger capacity, and 2.) if it is bigger, it will "appear as" the size of the image, until its reformatted at some point. If i need to copy an 8gb uSD stat and all i have is a 16gb usb stick, i'm not really concerned about the extra space, I probably just need the image duped asap.

**Comment by sunapi386** [comment]
Do we have an ETA on this?

**Comment by lurch** [comment]
This is a feature we'll need for https://etcher.io/pro so it'll probably get added sooner rather than later ;-)

**Comment by KerstinMaur** [comment]
[Etcher Pro](https://etcher.io/pro/) and [this feature](https://github.com/resin-io/etcher/issues/266) are awesome! 😍
Looking forward to them! 🧐

**Comment by mirceal** [comment]
+1

Any update on this?
Thanks

**Comment by jhermsmeier** [comment]
Sure, the technical foundation is pretty much there, and the user interface part is being worked on soon, and this will most likely end up in the next release :)

**Comment by MDM63** [comment]
Is this going to be a feature for Etcher or only for Etcher pro?

**Comment by jhermsmeier** [comment]
@MDM63 This is going to be available in both Etcher & Etcher Pro; generally both of them will maintain feature parity in regards to the writing process.

**Comment by cjcharles0** [comment]
I dont want to appear ungrateful, but is there an ETA on this yet? Such a useful feature (even if it doesnt scan the partition table to avoid copying blank space)

**Comment by TommySharpNZ** [comment]
+1 for this feature too please!

**Comment by kriekprinsloo** [comment]
+1 I also think that this would be great.

**Comment by inovek** [comment]
+1

**Comment by richardhughes260** [comment]
Is there an ETA on the backup feature?

**Comment by ddomit** [comment]
+1 Any news?

**Comment by dhowe** [comment]
progress?

**Comment by JaffaBoy** [comment]
hows this coming along?

**Comment by boelle** [comment]
yep... seems a shame that we can create an image from the sd card :-(

**Comment by Poincare1** [comment]
Any progress in being able to create an image from flash media?  Seems it was reported more than a year ago it would be an upcoming feature.

**Comment by sunapi386** [comment]
Clonezilla is a capable alternative, if you don't mind rebooting booting from USB and waiting. It's not bad if you dedicate a Raspberry Pi to do it. https://clonezilla.org/show-live-doc-content.php?topic=clonezilla-live/doc/01_Save_disk_image 

Clonezilla source code and see example bash scripts for cloning in "samples" folder. https://gitlab.com/stevenshiau/clonezilla.git

**Comment by PTS93** [comment]
@jhermsmeier is this feature put on the backlog for now? It sounded like it was close to release a year ago.

**Comment by jhermsmeier** [comment]
@PTS93 I don't know, I'm not working on Etcher anymore, but maybe @thundron or @zvin can answer that question

**Comment by paddywwoof** [comment]
Any news? @thundron  @zvin

**Comment by thundron** [comment]
@paddywwoof Working on it, but it will still need a bit more before it's pushed and merged, it requires some changes in the UI flows too

**Comment by rin67630** [comment]
By the way: to get rid of the useless data, you could implement a simple preliminary step: write a growing file with zeroes to the DS card until until it returns "full", then delete that file...
The compressed result should be perfect...

**Comment by mitra42** [comment]
Judging by the length of this topic, I guess this is never going to happen.  RPI's do a good job (mostly) with this with their SD Card copier, but seems to be incompatible with something (maybe usbmount)  - would be great to see this is in Balena Etcher

**Comment by jadonk** [comment]
I'd really like to see this pushed across the finish line.

**Comment by kevincastejon** [comment]
You should have listened to @alexandrosm 3 years ago when he was proposing to implement a "dumb" copying of the image as a start and improve it later..
I'm really waiting for this feature to come.

**Comment by rradar** [comment]
This "feature" is present in [usbimager](https://gitlab.com/bztsrc/usbimager/) btw :rocket:

**Comment by blinkenlight** [comment]
@rradar Thank you kind sir, your comment easily being the single most useful one on this ancient thread - which I can now finally leave behind - considering the fine gentlemen in charge of this here repo clearly put continental tectonic plates (and even some high-profile open source bug trackers with ongoing decade-old unresolved major issues) to shame in their eagerness to implement an obviously widely desired dead-simple feature. As to the fine gentlemen themselves: it's okay, you may censor this post as well if you wish; you've read it and that was the point - also, you are aware everyone who asked for this feature already got an email with the stuff you oh-so-promptly removed, right...?

**Comment by iKarith** [comment]
I'm no longer looking for this feature in balenaEtcher.

I'd suggest that garnering a reputation for deleting criticism isn't good PR, but doubt anyone at balena cares what I think.

**Comment by thundron** [comment]
@iKarith given it wasn't criticism but just banter and veiled insults for a completely different issue (not that would be justified in a different issue), yes, we don't want those to appear anywhere ideally and it's not the first time either.
Now on criticism - it can be hard, but we don't and will never disapprove it and. we even welcome it as much as we can - when it stays in the boundaries of criticism, that is.
Let's keep the discussion on the subject now - this issue is for backing up a drive and I'd like it to stay so :)

**Comment by shawaj** [comment]
For anyone just finding this you can now clone devices in balenaEtcher since v1.5.107

You can find the latest versions on the [balenaEtcher website](HTTPS://etcher.io) or from the [GitHub latest release page](https://github.com/balena-io/etcher/releases/latest)

**Comment by lurch** [comment]
> For anyone just finding this you can now clone devices in balenaEtcher since v1.5.107

But I guess there's still a slight difference between "clone" (i.e. copy the contents of one SD card to another SD card) and "backup" (i.e. create an image of your SD card on your hard drive) ? :wink:

**Comment by shawaj** [comment]
@lurch yep for sure! Have left it open as it is not really a fix - but does allow you to "backup" an image - at least in one sense of the word :-)

**Comment by daviderickson** [comment]
Etcher is so good at writing images, it would be great if the other half of the pipeline of creating images could be covered.  And in particular, as the size of cards continues to grow, having an ability to go straight to a compressed image with no intermediate uncompressed file would be tremendously useful.

**Comment by thundron** [comment]
@daviderickson that would mean having the image information *outside* of the compressed image, which I'm not sure is possible, or decompressing directly to the media drive which can be significantly slower than just decompressing and *then* flashing

**Comment by daviderickson** [comment]
> @daviderickson that would mean having the image information _outside_ of the compressed image, which I'm not sure is possible, or decompressing directly to the media drive which can be significantly slower than just decompressing and _then_ flashing

What image information are you referring to?  .img files are raw byte by byte transfers, there wouldn't be any need to decompress part of it to determine how to write it.  No question that writing a compressed image would likely be slower than an uncompressed one, but if it saved many GB of temporary disk usage and writes in the meantime, I'd gladly make the trade.

**Comment by charettepa** [comment]
any update on this, it seems to have been in the works since 2016

**Comment by doug62** [comment]
+1 This - The ability to copy a snapshot/image to a host iso so that they can be cloned later is critical (and so easy given the app's current features)

**Comment by LeeBinder** [comment]
How many hours of coding do you think this would take?

**Comment by djgranger** [comment]
6 years and still no movement.

**Comment by LeeBinder** [comment]
indeed, embarrassing.

**Comment by mattbland** [comment]
Just to add another voice that I'd like this feature. Also, compressing to .7z file rather than zip as much better compression. An option to suggest how big an image to make. When inserting a 64GB card with a 8Gb image on it, an possibly a previous image on the card, it might be an idea to have an option for 'erase card before proceeding?' when imaging new media. When creating backup then have an 'Advanced Option' to allow 'select image size' - in MB or GB from the start of the card/disk and discard the rest.

**Comment by lurch** [comment]
> When creating backup then have an 'Advanced Option' to allow 'select image size' - in MB or GB from the start of the card/disk and discard the rest.

AFAIK you can't just "discard the rest" because that would lead to filesystem corruption. I.e. you'd actually need to resize (shrink) the filesystem on the card before you're able to make a "smaller backup" of it. Have a look at e.g. https://github.com/Drewsif/PiShrink (which I think only runs on Linux, because Windows and Mac don't support Linux filesystems).

**Comment by matthew-l-weber** [comment]
I ended up putting a script together to fill this gap....

https://gist.github.com/matthew-l-weber/fe9b5b3a03305f2aee6d92849bc99f8b

**Comment by LeeBinder** [comment]
Thank you @matthew-l-weber 

Two questions:
1. Can the resulting image be restored with Balena Etcher?
2. When you write `error "Image file exists, skilling capture..."`, do you rather mean `error "Image file exists, killing capture..."`?

.. and one suggestion - instead:
```
#  -z         Compress image after shrinking with gzip
#  -Z         Compress image after shrinking with xz
```
-> 
```
#  -g         Compress image after shrinking with gzip
#  -x         Compress image after shrinking with xz
```

**Comment by matthew-l-weber** [comment]
> Thank you @matthew-l-weber
> 
> Two questions:
> 
>     1. Can the resulting image be restored with Balena Etcher?
> 
>     2. When you write `error "Image file exists, skilling capture..."`, do you rather mean `error "Image file exists, killing capture..."`?
> 
> 
> .. and one suggestion - instead:
> 
> ```
> #  -z         Compress image after shrinking with gzip
> #  -Z         Compress image after shrinking with xz
> ```
> 
> ->
> 
> ```
> #  -g         Compress image after shrinking with gzip
> #  -x         Compress image after shrinking with xz
> ```

1) Yes the resulting compressed image can be restored with Balena Etcher
2) I meant to have that error state that I'm skipping the capture as the image file already exists.  I'll move the code into the follow git repo so it's easier if others want to comment/pull request - https://github.com/matthew-l-weber/create-etcher-image/tree/main

**Comment by lurch** [comment]
>  I'll move the code into the follow git repo so it's easier if others want to comment/pull request - https://github.com/matthew-l-weber/create-etcher-image/tree/main

IMHO as your script makes use of PiShrink, it would only seem "polite" to credit https://github.com/Drewsif/PiShrink in https://github.com/matthew-l-weber/create-etcher-image/blob/main/README.md ?

**Comment by LeeBinder** [comment]
Mighty fine. xz is more efficient than gzip anyway.

However, I get 

`git@github.com: Permission denied (publickey). fatal: could not read from remote repository`

I can git from other repositories no problem, though. Do you need to make it available for the public?

**Comment by matthew-l-weber** [comment]
> Mighty fine. xz is more efficient than gzip anyway.
> 
> However, I get
> 
> `git@github.com: Permission denied (publickey). fatal: could not read from remote repository`
> 
> I can git from other repositories no problem, though. Do you need to make it available for the public?

It was public but I noticed I had used ssh cloning.  I've switched it to http so things should be good.

**Comment by KaKi87** [comment]
Hello,

Any news on this ?

If Etcher can image a drive and clone it to another, it certainly could just export said image to a file.

Thanks

**Comment by pschonmann** [comment]
I give up and going to backup with https://github.com/stevenshiau/clonezilla

**Comment by meta-vi** [comment]
> I give up and going to backup with https://github.com/stevenshiau/clonezilla

Hello, are you using this tool to cloning Tinkerboard image? Could you give me some guide to use it.
Thanks

**Comment by enessmr** [comment]
just turn on RAID-5 and copy all of the data of ur disk then bam you can flash image now

**Comment by rickdgray** [comment]
So this feature hasn't had any movement?

**Comment by JohnHind** [comment]
Just to had my voice to this, I was incredibly disappointed that Etcher allows both image files and drives as source, but only drives as destination. I'd want it to write .ISO files, but a choice between .IMG and .ISO would be even better. I cannot find any way to make ISO files on Mac since MacOS Disk utility does not support this format and in any case seems to be broken for restoring full disk images. Sadly, full disk images in .IMG format made in Disk Utility do not seem to be writable using Etcher either.
